### PR TITLE
Lucien/goal in image

### PR DESCRIPTION
**Pull request by LucienMorey**

### CODE REVIEW

File: goal_map.py [hunk 1]
```diff
@@ -1,12 +1,63 @@
...
+        826.326 + 120.0,
+        522.288 + 120.0,
+        0.0 + 120.0,
+    ]
+
+    X_DISTANCE_CUBE_LOOK_UP: list[float] = [353.22, 796.777, 1197.227]
+    X_DISTANCE_CONE_LOOK_UP: list[float] = [364.231, 795.231, 1167.655]
+
+    def __init__(self, on_blue_alliance: bool):
+        self.map: list[GoalRegionState] = []
```
Comment (auscompgeek) [inline goal_map.py]: This might be easier to understand if this were an ndarray - the grid is a fixed size after all.

File: goal_map.py [hunk 2]
```diff
@@ -1,12 +1,63 @@
...
+
+        # This will start from the top row of goals
+        for row in range(3):
+            cone_height = self.Z_DISTANCE_CONE_LOOK_UP[row]
+            cube_height = self.Z_DISTANCE_CUBE_LOOK_UP[row]
+            # this will start from the grid closest to the field origin
+            for grid in range(3):
+                # this will start from the goal closest to the field origin in each grid
+                for goal in range(3):
```
Comment (auscompgeek) [inline goal_map.py]: The game manual calls each of them nodes, right? Probably clearer than goal.

File: goal_map.py [hunk 3]
```diff
@@ -1,12 +1,63 @@
...
+                    id = row * 9 + grid * 3 + goal
+                    expected_game_piece = ExpectedGamePiece.CONE
+                    position = Translation3d()
+                    position.y = self.Y_DISTANCE_BETWEEN_GOALS * (grid + goal)
+                    if not on_blue_alliance:
+                        position.y += self.INTAKE_ZONE_OFFSET
+                    position.z = cone_height
+                    position.x = self.X_DISTANCE_CONE_LOOK_UP[grid]
```
Comment (auscompgeek) [inline goal_map.py]: Translation3ds are immutable

**Review by auscompgeek — COMMENTED** [review]
I know this is a draft, but here are some initial thoughts.

File: vision.py [hunk 4]
```diff
@@ -182,6 +186,65 @@ def annotate_image(
...
+    )
+
+
+def point2d_in_image_frame(pixel: np.ndarray, frame: np.ndarray):
+    return (
+        (pixel[0] >= 0)
+        and (pixel[0] < frame.shape[1])
+        and (pixel[1] >= 0)
+        and (pixel[1] < frame.shape[0])
+    )
```
Comment (LucienMorey) [inline vision.py]: @auscompgeek this ends up creating a ```numpy.bool_``` type. Do you know why this is? it becomes an issue in my unit tests where I try to assert do something like ```assert result is  True```. I believe I can just do ```assert result``` instead but this seems like a typing failure waiting to cause problems in the code

File: goal_map.py (hunk 3 above)
Comment (LucienMorey) [reply goal_map.py]: I had a quick check to see if there was anything else that seemed more appropriate in the docs, but it doesn't look like it. I think this means just keep track of the variables and create an object as the last step. Is this your perspective as well?

Comment (auscompgeek) [reply goal_map.py]: I suppose the other alternative is to add Translation3ds together

File: camera_config.py [hunk 5]
```diff
@@ -11,3 +12,9 @@ class CameraParams:
     transform: Pose3d  # pose of the camera w.r.t to the base link of the robot
     K: np.ndarray  # [3 x 3] intrinsic camera matrix
     fps: int
+
+    def get_horizontal_fov(self) -> float:
+        return 2 * atan2(self.width / 2, self.K[0][0])
```
Comment (auscompgeek) [inline camera_config.py]:
```suggestion
        return 2 * atan2(self.width / 2, self.K[0, 0])
```

File: camera_config.py [hunk 6]
```diff
@@ -11,3 +12,9 @@ class CameraParams:
...
+
+    def get_vertical_fov(self) -> float:
+        return 2 * atan2(self.height / 2, self.K[1][1])
```
Comment (auscompgeek) [inline camera_config.py]:
```suggestion
        return 2 * atan2(self.height / 2, self.K[1, 1])
```

File: vision.py [hunk 7]
```diff
@@ -120,68 +124,122 @@ def is_game_piece_present(
...
+        bool: if point is visible
+    """
+    vertical_angle = atan2(point.z, point.x)
+    horizontal_angle = atan2(point.y, point.x)
+
+    return (
+        (point.x > 0)
+        and (
+            vertical_angle > -camera_params.get_vertical_fov() / 2
+            and vertical_angle < camera_params.get_vertical_fov() / 2
```
Comment (auscompgeek) [inline vision.py]:
```suggestion
            -camera_params.get_vertical_fov() / 2
            < vertical_angle < camera_params.get_vertical_fov() / 2
```

It'd probably be worth making a variable for the half FOV.

File: vision.py [hunk 8]
```diff
@@ -120,68 +124,113 @@ def is_game_piece_present(
...
+    world_to_robot = Transform3d(Pose3d(), Pose3d(robot_pose))
+    world_to_camera = world_to_robot + camera_params.transform
+    node_region_camera_frame = world_to_camera.inverse() + Transform3d(
+        node_region.position, Rotation3d()
+    )
+
+    # Check the robot is facing the right direction for the point by checking it is inside the FOV
+    return point3d_in_field_of_view(
+        node_region_camera_frame.translation(), camera_params
+    )
```
Comment (LucienMorey) [inline vision.py]: I guess this should probably get its own tests and things

File: vision.py [hunk 9]
```diff
@@ -120,68 +124,113 @@ def is_game_piece_present(
...
+
+    return (
+        (point.x > 0)
+        and (
+            vertical_angle > -camera_params.get_vertical_fov() / 2
+            and vertical_angle < camera_params.get_vertical_fov() / 2
+        )
+        and (
+            horizontal_angle > -camera_params.get_horizontal_fov() / 2
+            and horizontal_angle < camera_params.get_horizontal_fov() / 2
```
Comment (auscompgeek) [inline vision.py]:
```suggestion
            -camera_params.get_horizontal_fov() / 2
            < horizontal_angle < camera_params.get_horizontal_fov() / 2
```

File: test_images.py [hunk 10]
```diff
@@ -23,7 +27,32 @@ def read_test_data_csv(fname: str):
...
+            result.append(
+                (
+                    node_region_visible == "True",
+                    float(robot_x),
+                    float(robot_y),
+                    float(heading),
+                    int(node_region_id),
+                )
+            )
+    return result
```
Comment (auscompgeek) [inline test_images.py:49]: nit: we could return a list comprehension here instead
```suggestion
        return [
            (
                node_region_visible == "True",
                float(robot_x),
                float(robot_y),
                float(heading),
                int(node_region_id),
            )
            for (
                node_region_visible,
                robot_x,
                robot_y,
                heading,
                node_region_id,
            )
            in csv.reader(f)
        ]
```

**Review by auscompgeek — APPROVED** [review]
LGTM

**Pull request merged**